Checkstyle: Enable additional naming rules

ssoloff opened the pull request:
For whatever reason, the Google Checkstyle ruleset does not enable the ConstantName, LocalFinalVariableName, and StaticVariableName rules.  This PR enables those rules with the default formats.  However, I think this PR should be discussed a bit just in case the team feels one or more of these rules shouldn't be enabled or one of the default formats should be changed.

These rules are summarized below:

Rule | Applies to | Default format | Notes
:-- | :-- | :-- | :--
ConstantName | `static`, `final` fields | `^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$` |
LocalFinalVariableName | local, `final` variables, including `catch` parameters and resources in `try` statements | `^[a-z][a-zA-Z0-9]*$` | format same as LocalVariableName rule
StaticVariableName | `static`, non-`final` fields | `^[a-z][a-zA-Z0-9]*$` | format same as MemberName rule

Enabling these rules increases the Checkstyle violation counts as follows:

Rule | main | test | integTest | | Total
:-- | :-- | :-- | :-- | :-- | :--
ConstantName | +295 | +34 | 0 | | +329
StaticVariableName | +85 | +5 | 0 | | +90
LocalFinalVariableName | +186 | +40 | 0 | | +226
| | | | |
**Total** | +566 | +79 | 0 | | +645

DanVanAtta commented:
I suspect the final constant field rule is not on by google checkstyle since they reserve caps for final immutable fields, not just final.

ssoloff commented:
> I suspect the final constant field rule is not on by google checkstyle since they reserve caps for final immutable fields, not just final.

:+1:  Would it be reasonable then to use a format that allowed either `UPPER_SNAKE_CASE` or `camelCase` for `static`, `final` fields as opposed to just leaving it unspecified?

DanVanAtta commented:
That is a good suggestion @ssoloff

ssoloff commented:
I mistakenly identified the default LocalFinalVariableName and StaticVariableName rules as matching our existing LocalVariableName and MemberName rules, respectively.  There was a slight difference due to the Google rules requiring a minimum of two character names.  I updated the configuration so that these rules match accordingly.

I also updated the ConstantName format to match either MemberName or UPPER_SNAKE_CASE.  To be consistent with the other Google rules, I required UPPER_SNAKE_CASE matches to have a minimum of two characters (not including the underscore).

The latest rule configuration is summarized below:

Rule | Format | Notes
:-- | :-- | :--
ConstantName | `^([a-z][a-z0-9][a-zA-Z0-9]*\|[A-Z][A-Z0-9]+(_[A-Z0-9]+)*)$` | same as MemberName -or- UPPER_SNAKE_CASE
StaticVariableName | `^[a-z][a-z0-9][a-zA-Z0-9]*$` | same as MemberName
LocalFinalVariableName | `^[a-z]([a-z0-9][a-zA-Z0-9]*)?$` | same as LocalVariableName

The new Checkstyle violation count changes follow:

Rule | main | test | integTest | | Total
:-- | :-- | :-- | :-- | :-- | :--
ConstantName | +224 | 0 | 0 | | +224
StaticVariableName | +85 | +5 | 0 | | +90
LocalFinalVariableName | +426 | +51 | 0 | | +477
| | | | |
**Total** | +735 | +56 | 0 | | +791

ssoloff commented:
Rebased against `master` at 5e95f81 (1.9.0.0.5816).

RoiEXLab commented:
@ssoloff Make sure those chekstyle rules are ignoring serialVersionUIDs, we can't change those!

ssoloff commented:
@RoiEXLab Good point.  `serialVersionUID` does not violate the StaticVariableName rule added in this PR, so we're good.

The real problem for this variable name would be due to the AbbreviationAsWordInName rule.  However, the Google ruleset disables that rule for static variables, probably to avoid triggering on UPPER_SNAKE_CASE, as you pointed out in another PR.

ssoloff commented:
Rebased against `master` at e7dc24b (1.9.0.0.5855).

ron-murhammer commented:
Looks good to me. Merging.
